File: api/src/test/java/org/apache/gravitino/semantic/TestSemanticModelSupportingTypes.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.semantic;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotSame;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

public class TestSemanticModelSupportingTypes {

  @Test
  public void testBuildSupportingTypes() {
    DialectExpression dialectExpression =
        DialectExpression.builder()
            .withDialect(Dialects.ANSI_SQL)
            .withExpression("order_amount")
            .build();
    Expression expression =
        Expression.builder().withDialects(new DialectExpression[] {dialectExpression}).build();
    Dimension dimension = Dimension.builder().withIsTime(false).build();
    AIContextObject contextObject =
        AIContextObject.builder()
            .withInstructions("Use certified fields")
            .withSynonyms(new String[] {"amount"})
            .withExamples(new String[] {"Revenue by month"})
            .withAdditionalProperties(Map.of("priority", 1))
            .build();
    CustomExtension extension =
        CustomExtension.builder().withVendorName("example").withData("{\"key\":\"value\"}").build();

    assertArrayEquals(new DialectExpression[] {dialectExpression}, expression.dialects());
    assertEquals(Boolean.FALSE, dimension.isTime());
    assertEquals("Use certified fields", contextObject.instructions());
    assertArrayEquals(new String[] {"amount"}, contextObject.synonyms());
    assertArrayEquals(new String[] {"Revenue by month"}, contextObject.examples());
    assertEquals(BigInteger.ONE, contextObject.additionalProperties().get("priority"));
    assertEquals("example", extension.vendorName());
    assertEquals("{\"key\":\"value\"}", extension.data());
  }

  @Test
  public void testExpressionDefensivelyCopiesDialects() {
    DialectExpression ansi =
        DialectExpression.builder().withDialect(Dialects.ANSI_SQL).withExpression("id").build();
    DialectExpression[] dialects = {ansi};
    Expression expression = Expression.builder().withDialects(dialects).build();

    dialects[0] =
        DialectExpression.builder()
            .withDialect(Dialects.BIGQUERY)
            .withExpression("changed")
            .build();

    assertEquals(Dialects.ANSI_SQL, expression.dialects()[0].dialect());
    assertNotSame(expression.dialects(), expression.dialects());
  }

  @Test
  @SuppressWarnings("unchecked")
  public void testAIContextVariantsAndAdditionalPropertiesAreImmutable() {
    List<Object> nestedValues = new ArrayList<>();
    nestedValues.add("first");
    Map<String, Object> nested = new LinkedHashMap<>();
    nested.put("values", nestedValues);
    Map<String, Object> additionalProperties = new LinkedHashMap<>();
    additionalProperties.put("nested", nested);

    AIContextObject contextObject =
        AIContextObject.builder().withAdditionalProperties(additionalProperties).build();
    AIContext objectContext = AIContext.of(contextObject);
    AIContext textContext = AIContext.of("Use certified metrics");

    nestedValues.add("changed");
    nested.put("extra", true);
    additionalProperties.put("new", "value");

    assertTrue(textContext.isText());
    assertEquals("Use certified metrics", textContext.text());
    assertNull(textContext.object());
    assertFalse(objectContext.isText());
    assertNull(objectContext.text());
    assertEquals(contextObject, objectContext.object());

    Map<String, Object> storedNested =
        (Map<String, Object>) contextObject.additionalProperties().get("nested");
    assertEquals(List.of("first"), storedNested.get("values"));
    assertThrows(
        UnsupportedOperationException.class,
        () -> contextObject.additionalProperties().put("another", "value"));
    assertThrows(
        UnsupportedOperationException.class,
        () -> ((List<Object>) storedNested.get("values")).add("value"));
  }

  @Test
  public void testOptionalAIContextCollectionsPreserveNullAndEmpty() {
    AIContextObject unset = AIContextObject.builder().build();
    AIContextObject empty =
        AIContextObject.builder().withSynonyms(new String[0]).withExamples(new String[0]).build();

    assertNull(unset.synonyms());
    assertNull(unset.examples());
    assertArrayEquals(new String[0], empty.synonyms());
    assertArrayEquals(new String[0], empty.examples());
    assertNotEquals(unset, empty);
  }

  @Test
  @SuppressWarnings("unchecked")
  public void testAIContextNumbersAreCanonicalized() {
    Map<String, Object> nested = new LinkedHashMap<>();
    nested.put("short", (short) 2);
    nested.put("float", 0.25F);

    Map<String, Object> additionalProperties = new LinkedHashMap<>();
    additionalProperties.put("byte", (byte) 1);
    additionalProperties.put("integer", 2);
    additionalProperties.put("long", 3L);
    additionalProperties.put("bigInteger", new BigInteger("123456789012345678901234567890"));
    additionalProperties.put("double", 0.125D);
    additionalProperties.put("bigDecimal", new BigDecimal("0.123456789012345678901234567890"));
    additionalProperties.put("nested", nested);
    additionalProperties.put("array", new Number[] {4, 5L, 0.5F});

    AIContextObject contextObject =
        AIContextObject.builder().withAdditionalProperties(additionalProperties).build();
    Map<String, Object> values = contextObject.additionalProperties();

    assertEquals(BigInteger.ONE, values.get("byte"));
    assertEquals(BigInteger.valueOf(2), values.get("integer"));
    assertEquals(BigInteger.valueOf(3), values.get("long"));
    assertEquals(new BigInteger("123456789012345678901234567890"), values.get("bigInteger"));
    assertEquals(new BigDecimal("0.125"), values.get("double"));
    assertEquals(new BigDecimal("0.123456789012345678901234567890"), values.get("bigDecimal"));
    assertEquals(BigInteger.valueOf(2), ((Map<String, Object>) values.get("nested")).get("short"));
    assertEquals(new BigDecimal("0.25"), ((Map<String, Object>) values.get("nested")).get("float"));
    assertEquals(
        List.of(BigInteger.valueOf(4), BigInteger.valueOf(5), new BigDecimal("0.5")),
        values.get("array"));

    AIContextObject equivalentContext =
        AIContextObject.builder()
            .withAdditionalProperties(Map.of("number", Integer.valueOf(1)))
            .build();
    AIContextObject equivalentLongContext =
        AIContextObject.builder()
            .withAdditionalProperties(Map.of("number", Long.valueOf(1)))
            .build();
    assertEquals(equivalentContext, equivalentLongContext);
    assertEquals(equivalentContext.hashCode(), equivalentLongContext.hashCode());
  }

  @Test
  public void testAIContextAdditionalPropertyBounds() {
    Object maximumDepthValue = "value";
    for (int depth = 0; depth < AIContextObject.MAX_ADDITIONAL_PROPERTY_NESTING_DEPTH; depth++) {
      maximumDepthValue = Map.of("nested", maximumDepthValue);
    }
    AIContextObject.builder()
        .withAdditionalProperties(Map.of("maximumDepth", maximumDepthValue))
        .build();

    Object excessiveDepthValue = Map.of("nested", maximumDepthValue);
    assertThrows(
        IllegalArgumentException.class,
        () ->
            AIContextObject.builder()
                .withAdditionalProperties(Map.of("excessiveDepth", excessiveDepthValue))
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            AIContextObject.builder()
                .withAdditionalProperties(Map.of("unsupportedNumber", new AtomicInteger(1)))
                .build());
  }

  @Test
  public void testSupportingTypeValidation() {
    assertThrows(IllegalArgumentException.class, () -> DialectExpression.builder().build());
    assertThrows(IllegalArgumentException.class, () -> Expression.builder().build());
    assertEquals("", AIContext.of("").text());

    DialectExpression ansi =
        DialectExpression.builder().withDialect(Dialects.ANSI_SQL).withExpression("id").build();
    DialectExpression custom =
        DialectExpression.builder().withDialect("TRINO").withExpression("id").build();
    assertEquals("TRINO", custom.dialect());
    assertEquals("ANSI_SQL", Dialects.ANSI_SQL);
    assertThrows(
        IllegalArgumentException.class,
        () -> DialectExpression.builder().withDialect("").withExpression("id").build());
    assertThrows(
        IllegalArgumentException.class,
        () -> Expression.builder().withDialects(new DialectExpression[] {ansi, ansi}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> Expression.builder().withDialects(new DialectExpression[] {ansi, null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> AIContextObject.builder().withSynonyms(new String[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () -> AIContextObject.builder().withExamples(new String[] {null}).build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            AIContextObject.builder()
                .withAdditionalProperties(Map.of("instructions", "duplicate"))
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            AIContextObject.builder()
                .withAdditionalProperties(Map.of("unsupported", new Object()))
                .build());
    assertThrows(
        IllegalArgumentException.class,
        () ->
            AIContextObject.builder()
                .withAdditionalProperties(Map.of("not_finite", Double.NaN))
                .build());

    Map<String, Object> cyclic = new LinkedHashMap<>();
    cyclic.put("self", cyclic);
    assertThrows(
        IllegalArgumentException.class,
        () -> AIContextObject.builder().withAdditionalProperties(cyclic).build());
  }
}

File: clients/client-python/tests/unittests/api/semantic/test_semantic_model_supporting_types.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import unittest

from gravitino.api.semantic.ai_context import AIContext
from gravitino.api.semantic.ai_context_object import (
    MAX_ADDITIONAL_PROPERTY_NESTING_DEPTH,
    AIContextObject,
)
from gravitino.api.semantic.custom_extension import CustomExtension
from gravitino.api.semantic.data_type import DataType
from gravitino.api.semantic.dialect_expression import DialectExpression
from gravitino.api.semantic.dialects import Dialects
from gravitino.api.semantic.dimension import Dimension
from gravitino.api.semantic.expression import Expression


class TestSemanticModelSupportingTypes(unittest.TestCase):
    def test_data_type_uses_ossie_wire_values(self):
        self.assertEqual("Decimal", DataType.DECIMAL.value)
        self.assertEqual("DateTime", DataType.DATE_TIME.value)
        self.assertEqual("DateTimeTz", DataType.DATE_TIME_TZ.value)
        self.assertEqual(DataType.OPAQUE, DataType("Opaque"))

    def test_dialects_are_case_sensitive_constants(self):
        self.assertEqual("ANSI_SQL", Dialects.ANSI_SQL)
        self.assertEqual("BIGQUERY", Dialects.BIGQUERY)
        with self.assertRaises(TypeError):
            Dialects()

    def test_dialect_expression(self):
        dialect_expression = DialectExpression(Dialects.ANSI_SQL, "order_amount")
        equal_expression = DialectExpression(Dialects.ANSI_SQL, "order_amount")

        self.assertEqual(Dialects.ANSI_SQL, dialect_expression.dialect())
        self.assertEqual("order_amount", dialect_expression.expression())
        self.assertEqual(dialect_expression, equal_expression)
        self.assertEqual(hash(dialect_expression), hash(equal_expression))
        self.assertNotEqual(dialect_expression, "invalid")
        self.assertNotEqual(
            dialect_expression, DialectExpression("ansi_sql", "order_amount")
        )

    def test_dialect_expression_rejects_empty_fields(self):
        with self.assertRaisesRegex(ValueError, "dialect must not be null or empty"):
            DialectExpression("", "order_amount")
        with self.assertRaisesRegex(ValueError, "expression must not be null or empty"):
            DialectExpression(Dialects.ANSI_SQL, "")

    def test_expression_preserves_dialect_order(self):
        ansi = DialectExpression(Dialects.ANSI_SQL, "order_amount")
        snowflake = DialectExpression(Dialects.SNOWFLAKE, "ORDER_AMOUNT")
        expression = Expression([ansi, snowflake])

        self.assertEqual([ansi, snowflake], expression.dialects())
        self.assertEqual(Expression([ansi, snowflake]), expression)
        self.assertNotEqual(Expression([snowflake, ansi]), expression)

    def test_expression_returns_a_copy_of_its_dialects(self):
        ansi = DialectExpression(Dialects.ANSI_SQL, "order_amount")
        expression = Expression([ansi])

        expression.dialects().clear()

        self.assertEqual([ansi], expression.dialects())

    def test_expression_rejects_invalid_dialects(self):
        ansi = DialectExpression(Dialects.ANSI_SQL, "order_amount")

        with self.assertRaisesRegex(ValueError, "dialects must not be null or empty"):
            Expression([])
        with self.assertRaisesRegex(ValueError, r"dialects\[0\] must not be null"):
            Expression([None])
        with self.assertRaisesRegex(
            ValueError, "dialects must not contain duplicate dialect: ANSI_SQL"
        ):
            Expression([ansi, DialectExpression(Dialects.ANSI_SQL, "amount")])

    def test_dimension(self):
        self.assertIsNone(Dimension().is_time())
        self.assertTrue(Dimension(is_time=True).is_time())
        self.assertEqual(Dimension(True), Dimension(True))
        self.assertEqual(hash(Dimension(True)), hash(Dimension(True)))
        self.assertNotEqual(Dimension(True), Dimension(False))
        self.assertNotEqual(Dimension(True), "invalid")

    def test_custom_extension(self):
        extension = CustomExtension("acme", '{"unit": "usd"}')

        self.assertEqual("acme", extension.vendor_name())
        self.assertEqual('{"unit": "usd"}', extension.data())
        self.assertEqual(CustomExtension("acme", '{"unit": "usd"}'), extension)
        self.assertNotEqual(CustomExtension("other", '{"unit": "usd"}'), extension)
        self.assertNotEqual(extension, "invalid")

    def test_custom_extension_rejects_missing_fields(self):
        with self.assertRaisesRegex(ValueError, "vendorName must not be null"):
            CustomExtension(None, "data")
        with self.assertRaisesRegex(ValueError, "data must not be null"):
            CustomExtension("acme", None)

    def test_ai_context_holds_text(self):
        ai_context = AIContext.of("Use certified metrics only")

        self.assertTrue(ai_context.is_text())
        self.assertEqual("Use certified metrics only", ai_context.text())
        self.assertIsNone(ai_context.object())
        self.assertEqual(AIContext.of("Use certified metrics only"), ai_context)
        self.assertEqual(
            hash(AIContext.of("Use certified metrics only")), hash(ai_context)
        )
        self.assertNotEqual(ai_context, "invalid")

    def test_ai_context_holds_object(self):
        ai_context_object = AIContextObject(instructions="Use certified metrics only")
        ai_context = AIContext.of(ai_context_object)

        self.assertFalse(ai_context.is_text())
        self.assertIsNone(ai_context.text())
        self.assertEqual(ai_context_object, ai_context.object())

    def test_ai_context_rejects_unsupported_values(self):
        with self.assertRaisesRegex(
            ValueError, "AI context must be a string or an AIContextObject"
        ):
            AIContext.of(None)
        with self.assertRaisesRegex(
            ValueError, "AI context must be a string or an AIContextObject"
        ):
            AIContext.of(42)
        with self.assertRaisesRegex(
            ValueError, "AI context must contain exactly one of text or object"
        ):
            AIContext(None, None)
        with self.assertRaisesRegex(
            ValueError, "AI context must contain exactly one of text or object"
        ):
            AIContext("text", AIContextObject())

    def test_ai_context_object_defaults(self):
        ai_context_object = AIContextObject()

        self.assertIsNone(ai_context_object.instructions())
        self.assertIsNone(ai_context_object.synonyms())
        self.assertIsNone(ai_context_object.examples())
        self.assertEqual({}, ai_context_object.additional_properties())

    def test_ai_context_object_retains_additional_properties(self):
        ai_context_object = AIContextObject(
            instructions="Use certified metrics only",
            synonyms=["sales"],
            examples=["total revenue by month"],
            additional_properties={
                "audience": "finance",
                "thresholds": {"warn": 10, "limits": [1, 2.5, True, None]},
            },
        )

        self.assertEqual("Use certified metrics only", ai_context_object.instructions())
        self.assertEqual(["sales"], ai_context_object.synonyms())
        self.assertEqual(["total revenue by month"], ai_context_object.examples())
        self.assertEqual(
            {
                "audience": "finance",
                "thresholds": {"warn": 10, "limits": [1, 2.5, True, None]},
            },
            ai_context_object.additional_properties(),
        )

    def test_ai_context_object_returns_copies(self):
        synonyms = ["sales"]
        additional_properties = {"nested": {"key": "value"}}
        ai_context_object = AIContextObject(
            synonyms=synonyms, additional_properties=additional_properties
        )

        synonyms.append("revenue")
        additional_properties["nested"]["key"] = "mutated"
        ai_context_object.synonyms().clear()
        ai_context_object.additional_properties()["nested"]["key"] = "mutated"

        self.assertEqual(["sales"], ai_context_object.synonyms())
        self.assertEqual(
            {"nested": {"key": "value"}}, ai_context_object.additional_properties()
        )

    def test_ai_context_object_equality(self):
        ai_context_object = AIContextObject(
            instructions="instructions", additional_properties={"audience": "finance"}
        )
        equal_object = AIContextObject(
            instructions="instructions", additional_properties={"audience": "finance"}
        )

        self.assertEqual(equal_object, ai_context_object)
        self.assertEqual(hash(equal_object), hash(ai_context_object))
        self.assertNotEqual(AIContextObject(instructions="other"), ai_context_object)
        self.assertNotEqual(ai_context_object, "invalid")

    def test_ai_context_object_rejects_invalid_additional_properties(self):
        with self.assertRaisesRegex(
            ValueError,
            "additional property must not duplicate standard property: synonyms",
        ):
            AIContextObject(additional_properties={"synonyms": ["sales"]})

        with self.assertRaisesRegex(
            ValueError, "additional property name must be a string"
        ):
            AIContextObject(additional_properties={1: "value"})

        with self.assertRaisesRegex(
            ValueError,
            "Additional property audience has non-JSON-compatible value type: object",
        ):
            AIContextObject(additional_properties={"audience": object()})

        with self.assertRaisesRegex(
            ValueError, "Additional property ratio must contain a finite number"
        ):
            AIContextObject(additional_properties={"ratio": float("nan")})

        with self.assertRaisesRegex(
            ValueError,
            r"Additional property nested contains a map key that is not a string",
        ):
            AIContextObject(additional_properties={"nested": {1: "value"}})

    def test_ai_context_object_rejects_cyclic_additional_properties(self):
        cyclic = {}
        cyclic["self"] = cyclic

        with self.assertRaisesRegex(
            ValueError, r"Additional property nested\.self contains a cyclic value"
        ):
            AIContextObject(additional_properties={"nested": cyclic})

    def test_ai_context_object_rejects_deeply_nested_additional_properties(self):
        value = "leaf"
        for _ in range(MAX_ADDITIONAL_PROPERTY_NESTING_DEPTH + 1):
            value = [value]

        with self.assertRaisesRegex(ValueError, "exceeds maximum nesting depth of 100"):
            AIContextObject(additional_properties={"nested": value})

    def test_ai_context_object_rejects_none_elements(self):
        with self.assertRaisesRegex(ValueError, r"synonyms\[1\] must not be null"):
            AIContextObject(synonyms=["sales", None])
        with self.assertRaisesRegex(ValueError, r"examples\[0\] must not be null"):
            AIContextObject(examples=[None])
